firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/FirestoreMapDecoder.kt

```diff
@@ -0,0 +1,244 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.ktx.serialization
+
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.ktx.serialization.FirestoreAbstractDecoder.Constants.START_INDEX
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.SerialKind
+import kotlinx.serialization.descriptors.StructureKind
+import kotlinx.serialization.descriptors.elementNames
+import kotlinx.serialization.encoding.AbstractDecoder
+import kotlinx.serialization.encoding.CompositeDecoder
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.modules.EmptySerializersModule
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.serializer
+
+/**
+ * A skeleton implementation of a decoder that decodes a nested map (or list) of Firestore supported
+ * types as a @[Serializable] object (or its field).
+ *
+ * Most of the decode* methods have default implementation that delegates decodeValue(value: Any).
+ * The [decodeElementIndex] method need to be override in each of the subclasses to support decoding
+ * of a nested map or list.
+ *
+ * @param nestedObject The nested object (map or list) that that needs to be decoded.
+ * @param docRef The [DocumentReference] where this nested object is obtained from.
+ */
+private abstract class FirestoreAbstractDecoder(
+    private val nestedObject: Any,
+    private val docRef: DocumentReference
+) : AbstractDecoder() {
+
+    object Constants {
+        const val START_INDEX = 0
+    }
+    protected var elementIndex: Int = START_INDEX
+    protected var isCurrentDecodeElementNotNull: Boolean = true
+    protected lateinit var decodedElementDataType: SerialKind
+
+    /** A list of values that need to be decoded as fields of the [Serializable] object. */
+    abstract val decodeValueList: List<Any?>
+
+    /** Decodes an enum field by returning its index from the enum object's descriptor. */
+    final override fun decodeEnum(enumDescriptor: SerialDescriptor): Int {
+        val decodedEnumFieldName = decodeValueList.elementAt(elementIndex - 1)
+        // TODO: Add a EnumNamingProperties parameter, and convert decodedEnumFieldName based on it
+        // i.e. case insensitive, snake_case match camelCase, etc
+        val enumFieldNames = enumDescriptor.elementNames.toList()
+        return enumFieldNames.indexOf(decodedEnumFieldName)
+    }
+
+    /**
+     * Returns true if the current element being decoded is not null.
+     *
+     * In case this method returns false, [decodeNull] method will be used instead of [decodeValue]
+     * to decode this null value element.
+     */
+    final override fun decodeNotNullMark(): Boolean {
+        val result = isCurrentDecodeElementNotNull
```

Comment:
Using flags is brittle, since it's up to the developer to ensure it's updated, which is error prone. Could this value be computed?